### ISSUE TITLE
Code Execution Order

### ISSUE THREAD

**Issue by bgk-** [peartreegames/topiary#5]
Currently diverts simply jump to the associated bough, however if it's an inner bough it may skip over variable declarations.

Consider the following which would result in a `foo is undefined`:
```
=== OUTER {
  var foo = "bar"
  === INNER {
    foo = "baz"
  }
}
=> OUTER.INNER
```

I'm currently unsure the best way to resolve this, but this should be valid code and work as expected even if it's nonsensical. For memory sake I don't want to simply preload all variables (though this would be fine for globals).

**Comment by bgk-** [comment]
I _think_ I have a possible idea here.

I've added a precompiled pass to hoist variables within each scope so we can be sure that all variables will at least be declared before any Boughs. With that we can change the jump logic to instead have each section run all logic at the top of the scope before going into the next jump.

Terminology
 - jump: `=> OUTER.INNER`
 - target: `OUTER.INNER`
 - point: `OUTER` and `INNER`

So for the above, the jump would first go to the `OUTER` point and execute the block until it hit a Bough, then it would go to the next point. Since it's the final point in the target it just executes as normal.

I'm not sure how this will play out entirely for more complex stories but we'll see.

**Comment by bgk-** [comment]
I've removed the precompile variable hoisting, I think it's better to just let the user write out the story as is and have the code execute as is, but the jump logic described above can still work and will need to be implemented.

**Comment by bgk-** [comment]
Implemented with 8172461
